### PR TITLE
Clean up the JSObjectRef API

### LINKED ISSUE #27: Separate namespaces for methods and properties?
Over time, it’s likely that more and more methods will be added to JSObjectRefs. Currently, there are `get`, `set`, `jsValue`, `subscript`, and (shortly) `instanceof`. However, at least some of these methods are likely to appear in regular JS objects (specifically `get` and `set`). While there is the escape hatch of calling `object[dynamicMember: "get"]` or `object.get("get")`, this is less than optimal and it would be great if there was a way to have a separate namespace for the JavaScriptKit-defined  methods so they’d never clash with something from JS.

Maybe it would be sufficient to just add a regular `subscript` method `object["get"]` or `object[1]`?

### PR DESCRIPTION
**Pull request by j-f1**
Fixes #27.

- `get`/`set` → `dynamicMember` and subscripts
- `instanceof` (kept)
- `jsValue` (kept)
- `apply` → `callAsFunction`

### CODE REVIEW

**Comment by j-f1** [comment]
Inspired by the `ExpressibleByNilLiteral` protocol, I came up with this subscript-based `JSValueConvertible` protocol:

```swift
public protocol JSValueConvertible {
    subscript(jsValue _: ()) -> JSValue { get }
}

extension JSValue {
    public init(from convertible: JSValueConvertible) {
        self = convertible[jsValue: ()]
    }
}
```

**Comment by j-f1** [comment]
Alright, the tests are still failing because it seems like I can’t have `Array` conform to `JSValueConvertible` when either `Element: JSValueConvertible` or `Element == JSValueConvertible` :(

Any suggestions?

**Comment by j-f1** [comment]
Update: I’ve replaced all the function call/new logic with this pattern:

```swift
// JS version in comment
func() // func.apply(null, [])
func(arg1, arg2, arg3) // func.apply(null, [arg1, arg2, arg3])
func(args: [arg1, arg2, arg3]) // func.apply(null, [arg1, arg2, arg3])
func(this: foo) // func.apply(foo, [])
func(this: foo, arg1, arg2, arg3) // func.apply(foo, [arg1, arg2, arg3])
func(this: foo, args: [arg1, arg2, arg3]) // func.apply(foo,  [arg1, arg2, arg3])
func(new: arg1, arg2, arg3) // new func(arg1, arg2, arg3)
func(.new, args: [arg1, arg2, arg3]) // new func(arg1, arg2, arg3)
func(.new) // new func()
```

The last two I’m a little uncertain of.

**Comment by j-f1** [comment]
One last change I was going to make was changing the `instanceof` function to an operator. I tried it out with `<<` which isn’t super clear.

It seems Swift doesn’t allow letters in operators, so `instanceof` is out. There are a ton of Unicode characters allowed but the only ASCII ones are `!%&*+-/<=>?^|~`. Maybe `<<?` or `<==`?

**Comment by kateinoigakukun** [comment]
Sorry, I didn't take a look much now, but could you ignore .swiftpm directory?

**Comment by kateinoigakukun** [comment]
I agree with deprecating get and set and provide subscript. 

But on second thought, `new` and `instanceof` are not natural and not swifty interface.

This library provides JavaScript like APIs but also they should be swifty.

I think `instanceof`, `new` and `jsValue` are not common naming as members and if users want to use those names, they can use subscript accessor instead.

I disagree that APIs become unnatural caring about edge cases.

**Comment by j-f1** [comment]
Good point! I reverted those changes, so now the last few samples look like this:

```js
foo.new(arg1, arg2, arg3)
new foo(arg1, arg2, arg3)

foo.new(args: [arg1, arg2, arg3])
new foo(...[arg1, arg2, arg3])

foo.new()
new foo()
```

Additionally, I changed the JSValue code back to using `.jsValue()` and changed `instanceof` to add the method `isInstanceOf`.

File: IntegrationTests/TestSuites/Sources/PrimaryTests/main.swift [hunk 1]
```diff
@@ -296,9 +296,9 @@ Object_Conversion: do {
     try expectEqual(jsArray2[2], .boolean(false))
     _ = jsArray2.push!(5)
     try expectEqual(jsArray2.length, .number(4))
-    _ = jsArray2.push!(jsArray1)
+//    _ = jsArray2.push!(jsArray1)
 
-    try expectEqual(jsArray2[4], .object(jsArray1))
+//    try expectEqual(jsArray2[4], .object(jsArray1))
```
Comment (kateinoigakukun) [inline IntegrationTests/TestSuites/Sources/PrimaryTests/main.swift:301]: Could you revert these comments out?

File: Sources/JavaScriptKit/JSFunction.swift [hunk 2]
```diff
@@ -1,45 +1,44 @@
 import _CJavaScriptKit
 
-@dynamicCallable
 public class JSFunctionRef: JSObjectRef {
     @discardableResult
-    public func dynamicallyCall(withArguments arguments: [JSValueConvertible]) -> JSValue {
-        let result = arguments.withRawJSValues { rawValues -> RawJSValue in
+    func callAsFunction(this: JSObjectRef? = nil, args: [JSValueConvertible]) -> JSValue {
```
Comment (kateinoigakukun) [inline Sources/JavaScriptKit/JSFunction.swift:5]:
```suggestion
    public func callAsFunction(this: JSObjectRef? = nil, args: [JSValueConvertible]) -> JSValue {
```

File: Sources/JavaScriptKit/JSFunction.swift [hunk 3]
```diff
@@ -1,45 +1,44 @@
...
+        new(args: args)
     }
 
-    public func new(_ arguments: JSValueConvertible...) -> JSObjectRef {
-        return arguments.withRawJSValues { rawValues in
+    // Guaranteed to return an object because either:
+    // a) the constructor explicitly returns an object, or
+    // b) the constructor returns nothing, which causes JS to return the `this` value, or
+    // c) the constructor returns undefined, null or a non-object, in which case JS also returns `this`.
+    public func new(args: [JSValueConvertible]) -> JSObjectRef {
```
Comment (kateinoigakukun) [inline Sources/JavaScriptKit/JSFunction.swift:40]:
```suggestion
    public func new(arguments: [JSValueConvertible]) -> JSObjectRef {
```

I think there shouldn't be abbreviations in public interface. 🙏

**Pull request merged**